src/box/applier.cc

```diff
@@ -1252,6 +1341,563 @@ applier_on_rollback(struct trigger *trigger, void *event)
+	{applier_thread_return_batch, NULL},
+};
+
+static inline int
+applier_handle_raft(struct applier *applier, struct applier_tx_row *txr)
+{
+	if (unlikely(iproto_type_is_raft_request(txr->row.type))) {
+		return applier_handle_raft_request(applier, &txr->req.raft.req);
+	}
+	return 0;
+}
+
+/**
+ * The tx part of applier-in-thread machinery. Apply all the parsed
+ * transactions.
+ */
+static void
+applier_process_batch(struct cmsg *base)
+{
+	struct applier_data_msg *msg = (struct applier_data_msg *)base;
+	struct applier *applier = msg->base.applier;
+	struct applier_tx *tx;
+	stailq_foreach_entry(tx, &msg->txs, next) {
+		struct applier_tx_row *txr =
+			stailq_first_entry(&tx->rows, struct applier_tx_row,
+					    next);
+		raft_process_heartbeat(box_raft(), applier->instance_id);
+		if (txr->row.lsn == 0) {
+			if (applier_handle_raft(applier, txr) != 0)
+				diag_raise();
+			applier_signal_ack(applier);
+		} else if (applier_apply_tx(applier, &tx->rows) != 0) {
+			diag_raise();
+		}
+		if (applier->state == APPLIER_FINAL_JOIN &&
+		    instance_id != REPLICA_ID_NIL) {
+			say_info("final data received");
+			applier_set_state(applier, APPLIER_JOINED);
+			applier_set_state(applier, APPLIER_READY);
+			applier_set_state(applier, APPLIER_FOLLOW);
+		}
+	}
+
+	/* Return the message to applier thread. */
+	cmsg_init(&msg->base.base, return_route);
+	cpipe_push(&applier->applier_thread->thread_pipe, &msg->base.base);
+}
+
+/** The callback invoked on the message return to applier thread. */
+static void
+applier_thread_return_batch(struct cmsg *base)
+{
+	struct applier_data_msg *msg = (struct applier_data_msg *) base;
+	struct applier *applier = msg->base.applier;
+	struct fiber *reader = applier->thread.reader;
+	lsregion_gc(&applier->thread.lsr, msg->lsr_id);
+	stailq_create(&msg->txs);
+	msg->tx_cnt = 0;
+	if (!fiber_is_dead(reader))
+		fiber_wakeup(reader);
+}
+
+static inline void
+ibuf_move_tail(struct ibuf *src, struct ibuf *dst, size_t size)
+{
+	assert(size > 0);
+	void *ptr = ibuf_alloc(dst, size);
+	if (ptr == NULL) {
+		panic("Applier failed to allocate memory for incoming "
+		      "transactions on ibuf");
+	}
+	memcpy(ptr, src->wpos - size, size);
+}
+
+/** Get the next message ready for push. */
+static struct applier_data_msg *
+applier_thread_next_msg(struct applier *applier)
+{
+	struct applier_thread *thread = container_of(cord(), typeof(*thread),
+						     cord);
+	int cur = applier->thread.msg_ptr;
+	for (int i = 0; i < 2; i++) {
+		struct applier_data_msg *msg =
+			&applier->thread.msgs[(cur + i) % 2];
+		/*
+		 * Choose a message which's either unused or already staged for
+		 * pushing but not yet pushed to tx thread.
+		 */
+		if (msg->tx_cnt != 0) {
+			if (msg->tx_cnt > APPLIER_THREAD_TX_MAX)
+				continue;
+			if (stailq_empty(&thread->tx_pipe.input))
+				continue;
+			if (stailq_first_entry(&thread->tx_pipe.input,
+					       struct cmsg, fifo) !=
+			    &msg->base.base) {
```

Comment:
IMO better add a flag to the message struct indicating that the message was pushed rather than check cpipe internals.

Reply:
I did this:
```diff
diff --git a/src/box/applier.cc b/src/box/applier.cc
index 5421fb686..f6ac4c1af 100644
--- a/src/box/applier.cc
+++ b/src/box/applier.cc
@@ -1523,9 +1523,7 @@ applier_thread_push_tx(struct applier_thread *thread,
        stailq_add_tail_entry(&msg->txs, tx, next);
        msg->tx_cnt++;
 
-       if (stailq_empty(&thread->tx_pipe.input) ||
-           stailq_first_entry(&thread->tx_pipe.input, struct cmsg, fifo) !=
-           &msg->base.base) {
+       if (msg->tx_cnt == 1) {
                cmsg_init(&msg->base.base, tx_route);
                cpipe_push(&thread->tx_pipe, &msg->base.base);
        } else {

```

It would be nice to have a `msg->is_flushed` flag as well. Then I could get rid of `tx_pipe.input` access in `applier_thread_next_msg()`.
But it turned out quite untrivial.

There is a `pipe->on_flush` trigger, which I could use. But then it would either set `is_flushed` for all cbus messages coming to applier (not necessarily `applier_data_msg`, or I would have to stash the messages which need the trigger somewhere. I don't like either of the options, so I didn't do that yet.